Adding code analysis tools

marcpage opened the pull request:
Added CheckStyle and PMD code/style analysis. Also cleaned up code to comply. Now TODO: comments get pulled out into warnings during the build.

Speewack commented on src/main/java/org/randywebb/wlts/beans/DetailedMember.java:
```diff
@@ -20,916 +20,773 @@
-	public void setZip(String zip) {
-		put("zip", zip);
-	}
+    /** For logging if needed during debugging. */
```
This should be a comment, not a javadoc comment (// for a single-line comment or /*...*/ for a block comment

marcpage replied:
Yep. I'll change that

Speewack replied:
Cool, there's a bunch of these throughout the code. I didn't comment on all of them.
I guess since it is a class-level variable (private static Logger), one could argue that it should be a Javadoc comment, and I could buy that argument.
Usually, I consider Javadoc as API documentation and want it on all public entities. I don't typically require it for private entities, though it is OK to do it.

Speewack commented on src/main/java/org/randywebb/wlts/beans/DetailedMember.java:
```diff
@@ -20,916 +20,773 @@
+        StringBuilder builder = new StringBuilder();
+
+        builder.append("DetailedMember [");
+        appendIfNotNull(builder, "name", getName());
```
Ok, this is a nice function call, I like t better than my old logic.
It seems to miss appending a comma between values though, and is thus not equivalent to what I had before. am I missing something?

marcpage replied:
appendIfNotNull appends a trailing comma. Behavior should not change.

marcpage replied:
It does leave a trailing comma. If you want I can clean that up. It will take a little bit of logic change, but pretty minor.

Speewack replied:
Glad to know it leaves a trailing comma. That's much cleaner code!
The trailing comma for getFormattedBirthDateFull() is OK as it is consistent with prior behavior.

marcpage replied:
There was a change in behavior for the last item. I've added logic to prefix with a comma unless it is the first non-null element being added.

Speewack commented on src/main/java/org/randywebb/wlts/beans/DetailedMinistered.java:
```diff
@@ -14,127 +14,141 @@
+        for (District district : districts) {
+            Household districtHousehold = null == district.getDistrictLeaderIndividualId()
+                        ? null
+                        : idToHousehold.get(district.getDistrictLeaderIndividualId());
```
These short line splits are killing me...

marcpage replied:
When I increase the line limit to 200, I'll fix these

Speewack commented on src/main/java/org/randywebb/wlts/Main.java:
```diff
@@ -30,27 +30,38 @@
 *
 */
-public class Main {
+public final class Main {
```
Why would we make this a final class? To prevent anyone extending it?

Speewack commented on src/main/java/org/randywebb/wlts/Main.java:
```diff
@@ -30,27 +30,38 @@
     private static Logger log = LoggerFactory.getLogger(Main.class);
 
+    /** default constructor. */
+    private Main() {
```
Why do we need to call out an empty default constructor?
I don't think we ever instantiate this class anyway as there's no reason to.

marcpage replied:
The private default constructor prevents instantiation. Methods without docs cause an error. I can make the missing javadoc a warning. Any other suggestions?

Speewack replied:
It makes sense that we wouldn't want to instantiate the Main class.

Speewack commented on src/main/java/org/randywebb/wlts/beans/Household.java:
```diff
@@ -18,21 +18,21 @@
-    private static double milesPerLat = 68.9;
-    /** estimated miles per degree of longitude for Pflugerville, TX */
-    private static double milesPerLon = 59.7;
+    /** estimated miles per degree of latitude for Pflugerville, TX. */
```
Hmm, this seems OK for Pflugerville, but should this be moved to an external config in case we want it to work for wards in other parts of the world?

marcpage replied:
Would moving this to the app config file be ok? Really these values would be "good enough" for how we are using distance for anywhere in the continental US, but I'm good moving them out.

Speewack replied:
It is OK to leave them in there if they are "good enough." I don't plan on wide distribution of this app, but I've been conditioned to think about global applications and when I saw "for Pflugerville, TX" in the comment, it raised a flag.

We currently bundle the AppConfig properties into the jar file, so they really aren't user settings. To do this "right" we'd want to consider having a second config file for user settings that we can distribute outside of the war. I don't know that we can justify doing that at the moment. Maybe the first step is to put these in the current app config properties, but add some comments in the file to delineate App Settings from User Settings so we can refactor it later if we need to expose those settings.
We could also just create a 2nd properties file user-settings.properties (or something), bundle it into the jar, and load it in AppConfig (we can load multiple properties files into a single properties object) as a cleaner way to delineate App vs User settings, and separate it out from the jar when it becomes needed / warranted.

Speewack commented on src/main/java/org/randywebb/wlts/beans/Household.java:
```diff
@@ -81,16 +81,16 @@ public Household(JSONObject definition) {
-        final double BIGGER_THAN_THE_EARTH = 100000000.0;
-        double min = BIGGER_THAN_THE_EARTH;
-        final double HUNDRETH_OF_A_MILE = 0.01;
+        final double biggerThanTheEarth = 100000000.0;
```
Constants should be named as you had them BIGGER_THAN_THE_EARTH

Speewack commented on src/main/java/org/randywebb/wlts/beans/Teacher.java:
```diff
@@ -9,7 +9,7 @@
 //import org.slf4j.LoggerFactory;
 
-/** Represents a minister (previously home or visiting teacher) */
+/** Represents a minister (previously home or visiting teacher). */
```
Hmm, is it worth refactoring this class to call it "Minister" now? probably not, but it wouldn't be too difficult to do.

marcpage replied:
I'm questioning if this API will break down entirely in August anyway. The API still calls it HT and VT

Speewack replied:
Good point. Let's not refactor now, but watch and see what happens to their API. We can adjust this code when we need to update for their likely API change.

Speewack commented on src/main/java/org/randywebb/wlts/beans/Teacher.java:
```diff
@@ -23,7 +23,7 @@
 
-    /** Default constructor */
+    /** Default constructor. */
     public Teacher() {
```
I don't usually bother defining a default constructor unless I'm going to add logic to it.

marcpage replied:
This needed a default constructor for some reason (I don't add default constructors without a reason either)

Speewack replied:
👍

Speewack commented on src/main/java/org/randywebb/wlts/ldstools/rest/LdsToolsClient.java:
```diff
@@ -56,20 +56,20 @@ public LdsToolsClient(String user, String password) throws AuthenticationExcepti
 
     /**
-     * Deliberately package scoped to enable other ldstools api components
+     * Deliberately package scoped to enable other ldstools api components.
```
Comment says it is deliberately package scoped, but line 64 shows that we changed it to public at some point. Update (remove) the comment accordingly?

Speewack commented on src/main/java/org/randywebb/wlts/ldstools/rest/LdsToolsClient.java:
```diff
@@ -78,6 +78,7 @@ public static CloseableHttpClient getHttpClient() {
         @throws AuthenticationException In the event username or password is incorrect
     */
     private void signIn(String user, String password) throws AuthenticationException {
+        final int httpOk = 200;
```
Meh, I guess I speak fluent HTTP Status Codes :) It's OK, though I'd name it as a Constant "HTTP_OK"

marcpage replied:
The linter catches "magic numbers" I understood 200 perfectly as well.

Speewack replied:
Fair enough. I agree that it is more readable.
You define this in 2 different methods in the code. I don't think we warrant a Constants class or anything though. I remember that the HTTP libraries have these constants already defined, but I don't recall where offhand.

Speewack commented on src/main/java/org/randywebb/wlts/reports/DetailedMemberListCSV.java:
```diff
@@ -36,15 +40,22 @@
 
-        // List<DetailedMember> members = processDetailMembers(Thread.currentThread().getContextClassLoader().getResourceAsStream("detailedmembership.json"));
-        // List<Household> households = processHouseholds(Thread.currentThread().getContextClassLoader().getResourceAsStream("membership.json"));
+        /*
```
Ugh, I really don't like this automated wrap. This isn't a commented block, but a couple of troubleshooting lines of code that I enable to run against a test file. Now it is harder to switch them on and off.

marcpage replied:
Ah, ok, now that I know that I can clean this up to make them on/off-able

Speewack replied:
It's not that big of a deal, sorry for the rant (I was tired). This was an example of why I don't like hard-enforced line length limits. It looked like an auto formatter had wrapped all the lines to a very short length.

Really, I shouldn't have that kind of quick switch in the comments. That's not clean code, but it was convenient for remembering how to do a quick test.
I'll eventually move that to the unit tests instead of a commented line that I can turn on and off.

Speewack commented on src/main/java/org/randywebb/wlts/util/AppConfig.java:
```diff
@@ -13,18 +13,21 @@
 *
 */
-public class AppConfig extends Properties {
+public final class AppConfig extends Properties {
```
Not sure why this needs to be final. I guess it's OK, but I don't particularly need to restrict anyone from extending it.
What am I missing?

marcpage replied:
I think it was because it follows the Singleton model (if I recall correctly). If we need to refactor for subclasses we can remove the final as part of the refactor.

Speewack replied:
Cool! I've learned something :)
I can't tell you how many singletons I've done without making the class final, but I can learn better!

marcpage commented on src/main/java/org/randywebb/wlts/beans/package-info.java:
```diff
@@ -0,0 +1,4 @@
+package org.randywebb.wlts.beans;
```
I was hesitant to add them. I scoured the web for best practices hoping the html files would be preferred. The html files seem to be deprecated and the package-info is the preferred way. I still need to add documentation for each package.

Speewack replied:
I don't really have a problem with it. It's a good way (and probably the preferred way) to add package-level documentation for javadoc.
Documenting the packages is just not something I've ever needed to do in practice.
But I really don't have any objection.

marcpage commented on pom.xml:
```diff
@@ -18,6 +18,32 @@
       <plugins>
         <!-- Configure maven-compiler plugin to use a specific java syntax
           and to generate a binary for a specific jave version. -->
+		 <plugin>
```
We are not using the default config for CheckStyle. Let me know what you want to change and I'll update the config to your preferences.

marcpage commented on src/main/java/org/randywebb/wlts/Main.java:
```diff
@@ -93,13 +93,13 @@ public static void main(String... args) throws Exception {
 		final boolean verb = ministers || map || routes || wlts || ministered;
-		final boolean hasUsername = (verb && arguments.length > 0) || (!verb && arguments.length > 1);
-		final boolean hasPassword = hasUsername && ((verb && arguments.length > 1) || (!verb && arguments.length > 2));
+		final boolean hasUsername = verb && arguments.length > 0 || !verb && arguments.length > 1;
```
I agree. I will disable the extra parenthesis error and add them back in. I'm not all that familiar with best practices in Java. This config professes it is the official Java best practices, and I was questioning the parenthesis. I'll change any of these to comply with what you want over any supposed standard.

marcpage replied:
I'll add parenthesis and remove the check

marcpage commented on src/main/java/org/randywebb/wlts/beans/Assignment.java:
```diff
@@ -4,20 +4,17 @@
 
 	/** In case we want to log something */
-	private static Logger log = LoggerFactory.getLogger(Assignment.class);
+	//private static Logger log = LoggerFactory.getLogger(Assignment.class);
```
I commented out the loggers in any class that did not use them. I left them in there so they would be easy to add in once some logging is needed. I could add a trace in the constructor or something minimal in each class to make sure it gets used. Or I could investigate if I could disable the error for unused and see if I can have it ignore patterns.

Speewack replied:
It is OK to comment them out. I think you have the right idea.
An unused private entity will trigger a compile-time warning, so I see the justification in commenting them out.
I think it is better to leave them commented out than to create a log statement just to remove the compile warning.
Best is to add logging where it can be helpful, but not to artificially add things just to justify having the Logger instance active.

marcpage commented on src/main/java/org/randywebb/wlts/ldstools/rest/LdsToolsClient.java:
```diff
@@ -104,7 +102,7 @@ private void signIn(String user, String password) throws AuthenticationException
 	/**
 	 * Signs out of the lds.org Session
-	 */
+	 * /
```
Signout is private and never called, so I block commented out the whole block, including the doc comment. Do you have a suggestion for a better way to handle this?

Speewack replied:
I was commenting on the extra space between the closing * and / of the block comment. I think the syntax is wrong and am surprised that the compiler allows it.
*/ vs * /

No concern over commenting it all out.

Speewack replied:
Yep, that occurred to me while I was at lunch. It was one of those cartoon lightbulb moments.
I was only looking at the diffs, so I didn't see the full context, but I understand what you are saying now. You deliberately broke this comment ending so you could comment the whole block.

marcpage replied:
Do you want me to just remove the dead code instead of just comment it out?

Speewack replied:
Yes, you are welcome to remove the dead code. The beauty of SCM is that we can find it again if we ever need it for reference. No point in keeping it around.

marcpage commented on src/main/java/org/randywebb/wlts/beans/AbstractBean.java:
```diff
@@ -148,10 +149,10 @@ public AbstractBean update(JSONObject definition, String[] expectedKeys) {
-    public Map<String,Object> forCSV() {
-        HashMap<String,Object> results = new HashMap<String,Object>();
+    public Map<String, Object> forCSV() {
+        HashMap<String, Object> results = new HashMap<String, Object>();
```
This is actually required by the CSV module we are using. Since the "beans" are not strictly beans anymore, the other interface to use for CSV generation is Map<String,Object>. I wasn't happy with it either.

Speewack replied:
Perfect. I'm OK since it is required by an API that we're using. This may have even been my code, I can't remember. It showed up in the diff though, so I asked :)
Sounds like we're of the same mindset WRT collections and generics vs raw Objects, but sometimes we just have to use the Object due to external constraints.

marcpage commented on src/main/resources/checkstyle.xml:
```diff
@@ -0,0 +1,177 @@
+        <!-- Checks for Size Violations.                    -->
+        <!-- See http://checkstyle.sf.net/config_sizes.html -->
+        <module name="LineLength">
+            <property name="max" value="100"/>
```
Terrific! I will change this to 200. The original was 80 (the default) which was crazy small. I found (Google?, Oracle?) someone else set it to 100.

Speewack replied:
Thanks.
We set it to 200 for the NI spec because we have high-res screens these days. 200 is long for most things, but is occasionally useful for URLs and lines that I want to turn on / off :)
I think we landed on 200 because we were going to enforce it with code linting tools and editor configurations, so we wanted to be generous in the length.
In general, I don't expect us to have much code that needs to exceed 100 or 120 characters per line, but 200 has been very readable unless we're viewing in a small terminal window.

Speewack commented on pom.xml:
```diff
@@ -18,6 +18,32 @@
+				 <configLocation>checkstyle.xml</configLocation>
+				 <encoding>UTF-8</encoding>
+				 <consoleOutput>true</consoleOutput>
+				 <failsOnError>true</failsOnError>
```
I'm willing to give it a shot. 
We can always change it to warn if it becomes a problem.
My main challenge will be that these rules are different from the NI rules that I've got my IDE configured to apply. 
I just need to create a project-specific configuration in my IDE to adopt these rules. Until I do that, I'll constantly clash with the rules at every save.

Speewack commented on src/main/java/org/randywebb/wlts/beans/AbstractBean.java:
```diff
@@ -74,17 +74,17 @@ See setFromJSON() for how subclasses can handle non-string values in the JSON
 
-            } catch(NoSuchMethodException e) {
+            } catch (NoSuchMethodException e) {
                 log.error("No constructor " + clazz.getName() + "(JSONObject)");
```
It's a common thing that I see and is usually unintentional. I think I see where you've used the second argument in most other places in the code, so I wasn't sure if this was an oversight or on purpose.
I don't require that we pass the exception in log statements, but wanted to confirm that this was intentional vs oversight.
FYI, I often see people do: log.error("Message" + e.toString()) which really doesn't work for them as e.toString() usually just writes out a memory address instead of the error message :)

marcpage commented on src/main/resources/checkstyle.xml:
```diff
@@ -0,0 +1,177 @@
+
+    - the Sun Code Conventions at http://java.sun.com/docs/codeconv/
+
+    - the Javadoc guidelines at
```
It caught several things that java did not catch. If they are redundant, it doesn't hurt anything other than extra processor cycles, right?

Speewack replied:
That's fine. This was more of an FYI about the javadoc doclint.
It is fine to have both until / unless there is a conflict between them.

marcpage commented on src/main/resources/checkstyle.xml:
```diff
@@ -0,0 +1,177 @@
+
+    <!-- Checks that a package-info.java file exists for each package.     -->
+    <!-- See http://checkstyle.sf.net/config_javadoc.html#JavadocPackage -->
+    <module name="JavadocPackage"/>
```
It ensures we have package level documentation (which I'm wanting to add). Do you want to ensure package level documentation or have that be completely optional? If just optional I can remove this. Would you prefer the .html file at that point?

Speewack replied:
First, when we add package-level documentation, the package-info.java file is fine. I don't think we need to use an html file instead.
Second, I'm not passionate about requiring package-level documentation as it is something I've not ever seen value in when consuming an API. I am perfectly fine with having it where it makes sense, but not convinced that we should require it for every package in the application.
I don't want to end up creating empty package-info.java files just to satisfy this rule. I'm happy to provide and maintain meaningful documentation on packages where it is useful, but don't think that every package warrants such documentation.

marcpage replied:
Ok, I'll remove the requirement. There is some package level documentation I want to add to the "beans" (about how they are not really beans) and some basic description in a couple of other places, but I think you are right, there are many of the packages that it doesn't make sense for.

marcpage commented on src/main/resources/checkstyle.xml:
```diff
@@ -0,0 +1,177 @@
+
+    <!-- Checks whether files end with a new line.                        -->
+    <!-- See http://checkstyle.sf.net/config_misc.html#NewlineAtEndOfFile -->
+    <module name="NewlineAtEndOfFile"/>
```
Good point, are you ok if I make it a warning? (my editor automatically adds a newline at the end, so I would never see it)

Speewack replied:
You can make it a warning, and we'll see how much it comes up.

marcpage commented on src/main/resources/checkstyle.xml:
```diff
@@ -0,0 +1,177 @@
+
+    <!-- Checks for Size Violations.                    -->
+    <!-- See http://checkstyle.sf.net/config_sizes.html -->
+    <module name="FileLength"/>
```
I've made it a warning. We don't hit it right now

marcpage commented:
Ok, I think I'm done cleaning up from my cleaning up and addressing all the issues. If I missed something feel free to point it out. Thanks for all the great reviews

Speewack commented on src/main/java/org/randywebb/wlts/beans/AbstractBean.java:
```diff
@@ -19,319 +19,331 @@
+                    try {
+                        toFill.add(constructor.newInstance(new Object[] {(JSONObject) o}));
+                    } catch (InvocationTargetException e) {
+                        if (log.isErrorEnabled()) {
```
I usually only surround trace, debug, and info messages with the if block.
The extra boolean check isn't really necessary here as the code will only execute when there is an exception. I guess you'd save a bit of performance if the log is set to FATAL instead of ERROR, but it's not something that impacts the "normal flow" of the program.
It is OK to have it, but generally not necessary for anything > INFO level

Speewack commented:
It is working for WLTS capabilities. Merging to main
